File: Frontend/T-Shirt/src/services/couponService.ts
import { api } from "./api";
import type {
  Coupon,
  CouponValidationResponse,
  CreateCouponPayload,
} from "../types/coupon";
import type { PageResponse } from "../types/product";

export const couponService = {
  validateCoupon: async (code: string): Promise<CouponValidationResponse> => {
    const response = await api.post("/coupons/validate", { code });
    return response.data;
  },

  getAllCoupons: async (page = 0, size = 10): Promise<PageResponse<Coupon>> => {
    const response = await api.get("/admin/coupons", { params: { page, size } });
    return response.data;
  },

  getCouponById: async (id: number): Promise<Coupon> => {
    const response = await api.get(`/admin/coupons/${id}`);
    return response.data;
  },

  createCoupon: async (payload: CreateCouponPayload): Promise<Coupon> => {
    const response = await api.post("/admin/coupons", payload);
    return response.data;
  },

  updateCoupon: async (id: number, payload: CreateCouponPayload): Promise<Coupon> => {
    const response = await api.put(`/admin/coupons/${id}`, payload);
    return response.data;
  },

  toggleCouponStatus: async (id: number, active: boolean): Promise<Coupon> => {
    const response = await api.patch(`/admin/coupons/${id}/status`, null, { params: { active } });
    return response.data;
  },

  deleteCoupon: async (id: number): Promise<void> => {
    await api.delete(`/admin/coupons/${id}`);
  },
};

File: Backend/tshirt-store-backend/src/main/java/com/tshirtstore/service/CouponService.java
package com.tshirtstore.service;

import com.tshirtstore.dto.coupon.*;
import com.tshirtstore.entity.Coupon;
import com.tshirtstore.entity.Order;
import com.tshirtstore.entity.User;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.math.BigDecimal;

public interface CouponService {
    CouponValidationResponse validateUserCartCoupon(String code);
    BigDecimal calculateDiscount(Coupon coupon, BigDecimal subtotal);
    void applyCouponAndRecordUsage(String code, User user, Order order, BigDecimal discountAmount);

    Page<CouponResponse> getAllCoupons(Pageable pageable);
    CouponResponse getCouponById(Long id);
    CouponResponse createCoupon(CreateCouponRequest request);
    CouponResponse updateCoupon(Long id, CreateCouponRequest request);
    CouponResponse toggleCouponStatus(Long id, boolean active);
    void deleteCoupon(Long id);
}
